### ISSUE TITLE
poor accuracy for RF vs. sklearn when column sampling

### ISSUE THREAD

**Issue by pseudotensor** [microsoft/LightGBM#2536]
I introduced a leak of the target, and LightGBM has issues seeing the leak when RF is used with moderate column sampling (0.5 or less).  But sklearn has no issues:

[leak_issue_new.zip](https://github.com/microsoft/LightGBM/files/3795482/leak_issue_new.zip)

There is only noise level changes in the leaky feature compared to the target.

```
LightGBM train R2: 0.664828 eval R2: 0.638859
Sklearn RF train R2: 0.999209 eval R2: 0.932433
```

No other parameters (bins, estimators, learning rate, etc.) help LightGBM.  gbdt does fine.

The problem with larger column sampling or gbdt is if there are multiple leaks, lightgbm highly prefers the first features by order provided in X.

There is no way apparently to tell lightgbm to use a different seed for each tree.  Even if random seed is changed, it highly prefers the first features by order, so the variable importance is never able to randomly see the other features as important.  One would have to build a tree with randomized order of features, which seems LightGBM should have done this.

**Comment by pseudotensor** [comment]
Code for people who just want to stare and not download zip:

```
import _pickle as pickle
import lightgbm as lgb

X, y, eval_set = pickle.load(open("leak_issue.pkl", "rb"))

# LightGBM
params = {'boosting_type': 'rf', 'colsample_bytree': 0.3,
          'subsample': 0.632, 'subsample_freq': 1,
          'objective': 'regression', 'num_class': 1}
model = lgb.LGBMRegressor(**params)
model.fit(X, y)

from sklearn.metrics import r2_score
preds_train = model.predict(X)
preds_eval = model.predict(eval_set[0][0])
r2_train = r2_score(y, preds_train)
r2 = r2_score(eval_set[0][1], preds_eval)
print("LightGBM train R2: %g eval R2: %g" % (r2_train, r2))

# Sklearn
from sklearn.ensemble import RandomForestRegressor
max_features = int(params['colsample_bytree'] * X.shape[1])
model_sklearn = RandomForestRegressor(max_features=max_features)
model_sklearn.fit(X, y)

preds_train = model_sklearn.predict(X)
preds_eval = model_sklearn.predict(eval_set[0][0])
r2_train = r2_score(y, preds_train)
r2 = r2_score(eval_set[0][1], preds_eval)
print("Sklearn RF train R2: %g eval R2: %g" % (r2_train, r2))
```

See how we set max_features for sklearn to match lightgbm.

**Comment by guolinke** [comment]
@pseudotensor could you provide the ncols?

**Comment by guolinke** [comment]
I think the sklearn's RF is colsample by node as well. And it seems more leaves than your lgb setting.

**Comment by pseudotensor** [comment]
In my case there is only 1 leaky feature, I was just discussing if there were many how that complicates things even more.

I'll try colsample_bynode, but you said it will easily fail if leaky feature is selected, so why is that good?

As I mentioned, I tried many changes to hyper parameters, including number of leaves.  That has no effect here.

**Comment by guolinke** [comment]
I don't know about your data. But I try larger colsample, or enable colsample by node, larger number of leaves, smaller min_data. All of them can significantly improve the performance. So I guess your data needs more over-fit.
I also check the model file of your code, it didn't always use the first feature.

**Comment by pseudotensor** [comment]
But how possible with 1000 random trees of depth 10 and large leaves that sampling columns by 0.3 doesn’t overfit either?  There are only like 10 columns!

**Comment by guolinke** [comment]
@pseudotensor sorry, what is `10 columns` from?  
And for RF model, I don't think using more trees can over-fit. 

LightGBM indeed will use the feature with the smaller index when the split gain is the same.
I don't think this is problem, since the split gain should be different for different features.

**Comment by guolinke** [comment]
you can also use this branch: https://github.com/microsoft/LightGBM/tree/rand-feature to enable the random feature selection when the gain is the same. 
However, I guess the results should be the same.

I insist that the `num_leaves=31` and `colsample_bytree=0.3` is too simple for RF model. 
Especially compared with sklearn with ultimate leaves and `col_sample_bynode=0.3`.  

A simple test:

```
>>> X, y, eval_set = pickle.load(open("leak_issue.pkl", "rb"))
>>>
>>> # LightGBM
... params = {'boosting_type': 'rf', 'colsample_bynode': 0.3,
...           'subsample': 0.632, 'subsample_freq': 1, 'num_leaves': 500, 'min_child_samples': 2,
...           'objective': 'regression', 'num_class': 1}
>>> model = lgb.LGBMRegressor(**params)
>>> model.fit(X, y)
LGBMRegressor(boosting_type='rf', class_weight=None, colsample_bynode=0.3,
       colsample_bytree=1.0, importance_type='split', learning_rate=0.1,
       max_depth=-1, min_child_samples=2, min_child_weight=0.001,
       min_split_gain=0.0, n_estimators=100, n_jobs=-1, num_class=1,
       num_leaves=500, objective='regression', random_state=None,
       reg_alpha=0.0, reg_lambda=0.0, silent=True, subsample=0.632,
       subsample_for_bin=200000, subsample_freq=1)
>>>
>>> from sklearn.metrics import r2_score
>>> preds_train = model.predict(X)
>>> preds_eval = model.predict(eval_set[0][0])
>>> r2_train = r2_score(y, preds_train)
>>> r2 = r2_score(eval_set[0][1], preds_eval)
>>> print("LightGBM train R2: %g eval R2: %g" % (r2_train, r2))
LightGBM train R2: 0.965195 eval R2: 0.914147
```

**Comment by pseudotensor** [comment]
I modified the script to have the additional parameters you added: `'num_leaves': 500, 'min_child_samples': 2,` and I didn't see the same results:

```
import _pickle as pickle
import lightgbm as lgb

X, y, eval_set = pickle.load(open("leak_issue.pkl", "rb"))

# LightGBM                                                                                                                                                                                                                   
params = {'boosting_type': 'rf', 'colsample_bytree': 0.3,
          'subsample': 0.632, 'subsample_freq': 1,
          'num_leaves': 500, 'min_child_samples': 2,
          'objective': 'regression', 'num_class': 1}
model = lgb.LGBMRegressor(**params)
model.fit(X, y)

from sklearn.metrics import r2_score
preds_train = model.predict(X)
preds_eval = model.predict(eval_set[0][0])
r2_train = r2_score(y, preds_train)
r2 = r2_score(eval_set[0][1], preds_eval)
print("LightGBM train R2: %g eval R2: %g" % (r2_train, r2))

# Sklearn                                                                                                                                                                                                                    
from sklearn.ensemble import RandomForestRegressor
max_features = int(params['colsample_bytree'] * X.shape[1])
model_sklearn = RandomForestRegressor(max_features=max_features)
model_sklearn.fit(X, y)

preds_train = model_sklearn.predict(X)
preds_eval = model_sklearn.predict(eval_set[0][0])
r2_train = r2_score(y, preds_train)
r2 = r2_score(eval_set[0][1], preds_eval)
print("Sklearn RF train R2: %g eval R2: %g" % (r2_train, r2))
```

```
LightGBM train R2: 0.676676 eval R2: 0.629589
Sklearn RF train R2: 0.99941 eval R2: 0.963498
j
```

**Comment by pseudotensor** [comment]
Perhaps you guys fixed something in last couple months, since I'm not using the latest version.

**Comment by guolinke** [comment]
@pseudotensor ```colsample_bynode``` is important in your case, as it will use more “different” features in each tree.

**Comment by StrikerRUS** [comment]
Can this be closed?

**Comment by pseudotensor** [comment]
Ok
